### PR TITLE
CLI - pvbackup create, schedule, sync, restore and delete

### PR DESCRIPTION
**Pull request by v-harihar**
**Describe what this PR does**
The backup and restore is based on **restic** utility.

**Create:**
1. Create the secret referring restic config file
2. Create the replication source
3. Save the details to relationship file

**Schedule:**
1. Set schedule on the replication source for
2. Scheduled backups

**Sync:**
1. Trigger single manual backup by setting manual trigger

**Restore**
1. Create the NS, pvc, secret if doesn't exist or use the same if exists
2. Pause ReplicationSource
3. Create ReplicationDestination and trigger the restore
4. Unpause the ReplicationSource
5. Save the details to relationship file

**Delete:**
1. Delete the replicationSource/replicationDestination relationship
5. Delete the secret
6. Delete the relationship file

**Related issues:**
#193

Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

### CODE REVIEW

File: kubectl-volsync/cmd/pvbackup_create.go [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+		rs  *volsyncv1alpha1.ReplicationSource
+		err error
+	)
+	klog.Infof("waiting for replication source to be available")
+	err = wait.PollImmediate(5*time.Second, defaultRsyncKeyTimeout, func() (bool, error) {
+		rs, err = prd.getReplicationSource(ctx, client)
+		if err != nil {
+			return false, err
+		}
+		// TODO: What should be the condition to break the wait ?
```
Comment (v-harihar) [inline kubectl-volsync/cmd/pvbackup_create.go:305]: Need to find a condition that is ideal to say replication Source is ready. Please help me with the proper condition

**Review by JohnStrunk — COMMENTED** [review]
I'm having a bit of trouble untangling the usage a bit... There are 5 sub commands here: create, delete, restore, schedule, and sync.
I assume that the intention is that a given pvbackup relationship would be either a backup or a restore, not both? (the pause code is what's confusing me). Assuming this to be true:
- Backup flow would be: create -> (optional schedule / sync calls) -> delete
- Restore flow: restore -> delete
We should probably rename "create" to "backup".

File: kubectl-volsync/cmd/pvbackup_delete.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	prs := pd.pr.data.Source
+	rs, err := prs.getReplicationSource(ctx, pd.client)
+	if err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			klog.Infof("backup source %s not found, ignore", prs.RSName)
+			return nil
+		}
+		return fmt.Errorf("failed to get source CR, %w", err)
+	}
+	err = pd.client.Delete(ctx, rs)
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_delete.go:136]: That was just a foolish stuff from me.

File: kubectl-volsync/cmd/pvbackup_sync.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+type pvBackupSync struct {
+	pr *pvBackupRelationship
+}
+
+var pvBackupSyncCmd = &cobra.Command{
+	Use:   "sync",
+	Short: i18n.T("Run single manual backup"),
+	Long: templates.LongDesc(i18n.T(`
+	This command triggers the single backup instance and waits till backup to complete
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_sync.go:41]: sure

File: kubectl-volsync/cmd/pvbackup_create.go [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+		return err
+	}
+
+	// Save the replication source details into relationship file
+	if err = pc.pr.Save(); err != nil {
+		return fmt.Errorf("unable to save relationship configuration: %w", err)
+	}
+
+	// Remove the configuration file saved
+	os.Remove(pc.resticConfig.filename)
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_create.go:238]: sure

File: kubectl-volsync/cmd/pvbackup_create.go [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+	// Build struct pvBackupRelationshipSource from struct pvBackupCreate
+	pc.pr.data.Source = pc.newPVBackupRelationshipSource()
+	if err != nil {
+		return err
+	}
+
+	// Add restic configurations into cluster
+	err = pc.ensureSecret(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to create secrete, %w", err)
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_create.go:217]: correcting

File: kubectl-volsync/cmd/pvbackup_create.go [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+}
+
+func parseCronSpec(cs string) (*string, error) {
+	parser := cron.NewParser(cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow | cron.Descriptor)
+	if _, err := parser.Parse(cs); err != nil {
+		return nil, err
+	}
+
+	return &cs, nil
+}
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_create.go:199]: sure

File: kubectl-volsync/cmd/common_utils.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+func deleteSecret(ctx context.Context, ns types.NamespacedName, cl client.Client) error {
+	secret := &corev1.Secret{}
+	err := cl.Get(ctx, ns, secret)
+	if err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			klog.Infof("secret %s not found, ignore", ns.Name)
+			return nil
+		}
+		return fmt.Errorf("secret %s not found, %w", ns.Name, err)
+	}
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/common_utils.go:45]: My bad.

File: kubectl-volsync/cmd/pvbackup_create.go [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+		return fmt.Errorf("failed to fetch the restic-config, err = %w", err)
+	}
+	resticConfig, err := parseResticConfig(resticConfigFile)
+	if err != nil {
+		return err
+	}
+	pc.resticConfig = *resticConfig
+
+	keys := []string{"RESTIC_REPOSITORY", "RESTIC_PASSWORD",
+		"AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY"}
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_create.go:153]: sure

**Comment by v-harihar** [comment]
> I'm having a bit of trouble untangling the usage a bit... There are 5 sub commands here: create, delete, restore, schedule, and sync. I assume that the intention is that a given pvbackup relationship would be either a backup or a restore, not both? (the pause code is what's confusing me). Assuming this to be true:
> 
> * Backup flow would be: create -> (optional schedule / sync calls) -> delete
> * Restore flow: restore -> delete
>   We should probably rename "create" to "backup".

I am thinking how to make `command (pv-backup)` and `option (create ==> backup)` look bit nice. If I rename `create` to `backup` then it will be `kubectl volsync pv-backup backup ...`. 

Any suggestion for the name of the command ? Even for `restore`, the command is `pv-backup` which is not complementing well.

File: kubectl-volsync/cmd/pvbackup_create.go [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+		value, ok := v.Get(key).(string)
+		if !ok {
+			klog.Info("interface conversion: interface is not string")
+			return nil, os.ErrInvalid
+		}
+
+		if value == "" {
+			klog.Info("value missing for the key = %s", key)
+			return nil, os.ErrInvalid
+		}
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_create.go]: sure

File: kubectl-volsync/cmd/pvbackup.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+	This set of commands is designed to configure the restic repository to provide
+	automatic backups and restore`)),
+}
+
+func init() {
+	rootCmd.AddCommand(pvBackupCmd)
+
+	pvBackupCmd.PersistentFlags().StringP("relationship", "r", "", "relationship name")
+	cobra.CheckErr(pvBackupCmd.MarkPersistentFlagRequired("relationship"))
+	cobra.CheckErr(viper.BindPFlag("relationship", pvBackupCmd.PersistentFlags().Lookup("relationship")))
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup.go:123]: blind copy resulted in this. removing it

File: kubectl-volsync/cmd/pvbackup_create.go [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+		return err
+	}
+
+	// Save the replication source details into relationship file
+	if err = pc.pr.Save(); err != nil {
+		return fmt.Errorf("unable to save relationship configuration: %w", err)
+	}
+
+	// Remove the configuration file saved
+	os.Remove(pc.resticConfig.name)
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_create.go]: not required as per discussion. removing it

File: kubectl-volsync/cmd/pvbackup_schedule.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	This command schedules/reschedules backup instances.
+
+	ex: # kubectl volsync pv-backup schedule --cronspec "*/30 * * * *" --relationship pvb1
+	`)),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		ps, err := newPVBackupSchedule(cmd)
+		if err != nil {
+			return err
+		}
+		pr, err := loadPVBackupRelationship(cmd)
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply kubectl-volsync/cmd/pvbackup_schedule.go:51]: Yes, below condition was missed

```
if pr.data.Source == nil {
			return fmt.Errorf("incompatible relationship, %w", os.ErrInvalid)
}
```

Updating the patch

**Comment by JohnStrunk** [comment]
> > I'm having a bit of trouble untangling the usage a bit... There are 5 sub commands here: create, delete, restore, schedule, and sync. I assume that the intention is that a given pvbackup relationship would be either a backup or a restore, not both? (the pause code is what's confusing me). Assuming this to be true:
> > 
> > * Backup flow would be: create -> (optional schedule / sync calls) -> delete
> > * Restore flow: restore -> delete
> >   We should probably rename "create" to "backup".
> 
> I am thinking how to make `command (pv-backup)` and `option (create ==> backup)` look bit nice. If I rename `create` to `backup` then it will be `kubectl volsync pv-backup backup ...`.
> 
> Any suggestion for the name of the command ? Even for `restore`, the command is `pv-backup` which is not complementing well.

Maybe my suggestion wasn't so good... We could stick with the create/restore. Would changing the top-level command make it read any better? You do "create" a backup and then "restore" it later.

```console
$ kubectl volsync backup create ...
$ kubectl volsync backup restore ...
```

The part I'm struggling with is that the other commands require you to "create" a relationship, and that's what we've used the create for. Whereas here, you don't need to issue a create before a restore (for a given relationship).
I don't know what to suggest.

**Pull request closed without merging**